File: src/test/java/io/lsdconsulting/lsd/distributed/interceptor/interceptor/LsdFeignLoggerInterceptorShould.java
package io.lsdconsulting.lsd.distributed.interceptor.interceptor;

import feign.Logger;
import feign.Request;
import feign.Response;
import io.lsdconsulting.lsd.distributed.access.model.InterceptedInteraction;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.RequestCaptor;
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.ResponseCaptor;
import org.apache.commons.lang3.RandomUtils;
import org.junit.jupiter.api.Test;

import java.io.IOException;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

class LsdFeignLoggerInterceptorShould {

    private final RequestCaptor requestCaptor = mock(RequestCaptor.class);
    private final ResponseCaptor responseCaptor = mock(ResponseCaptor.class);
    private final Request request = mock(Request.class);
    private final Response response = mock(Response.class);

    private final LsdFeignLoggerInterceptor underTest = new LsdFeignLoggerInterceptor(requestCaptor, responseCaptor);

    private final Logger.Level level = Logger.Level.BASIC;
    private final Long elapsedTime = RandomUtils.nextLong();

    @Test
    void logsRequest() {
        underTest.logRequest("configKey", level, request);

        verify(requestCaptor).captureRequestInteraction(request);
    }
   @Test
    void logAndRebufferResponse() throws IOException {
        given(responseCaptor.captureResponseInteraction(any(), eq(elapsedTime))).willReturn(InterceptedInteraction.builder().build());

        underTest.logAndRebufferResponse("configKey", level, response, elapsedTime);

        verify(responseCaptor).captureResponseInteraction(response, elapsedTime);
    }
}

File: feign/src/test/kotlin/io/lsdconsulting/lsd/distributed/interceptor/interceptor/LsdFeignLoggerInterceptorShould.kt
package io.lsdconsulting.lsd.distributed.interceptor.interceptor

import feign.Logger
import feign.Request
import feign.Request.HttpMethod.POST
import feign.Response
import io.github.krandom.KRandom
import io.github.krandom.KRandomParameters
import io.lsdconsulting.lsd.distributed.connector.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.interceptor.captor.FeignRequestCaptor
import io.lsdconsulting.lsd.distributed.interceptor.captor.FeignResponseCaptor
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.apache.commons.lang3.RandomStringUtils
import org.apache.commons.lang3.RandomUtils.secure
import org.junit.jupiter.api.Test
import kotlin.text.Charsets.UTF_8

internal class LsdFeignLoggerInterceptorShould {
    private val kRandom = KRandom(KRandomParameters().seed(System.currentTimeMillis()))
    private val feignRequestCaptor = mockk<FeignRequestCaptor>()
    private val feignResponseCaptor = mockk<FeignResponseCaptor>()
    private val request = mockk<Request>(relaxed = true)
    private val level = Logger.Level.BASIC
    private val elapsedTime = secure().randomLong()
    private val body = RandomStringUtils.secure().nextAlphanumeric(10)

    private val underTest = LsdFeignLoggerInterceptor(feignRequestCaptor, feignResponseCaptor)

    @Test
    fun `logs request`() {
        every { feignRequestCaptor.captureRequestInteraction(request) } returns kRandom.nextObject(InterceptedInteraction::class.java)
        underTest.logRequest("configKey", level, request)
        verify {  feignRequestCaptor.captureRequestInteraction(request) }
    }

    @Test
    fun `log and re-buffer response`() {
        val request = Request.create(POST, "url", mapOf(), null, UTF_8, null)
        val response = Response.builder().body(body.byteInputStream(UTF_8), body.length).request(request).build()
        every { feignResponseCaptor.captureResponseInteraction(response, body, elapsedTime) } returns
            kRandom.nextObject(InterceptedInteraction::class.java).copy(body = null)

        underTest.logAndRebufferResponse("configKey", level, response, elapsedTime)

        verify { feignResponseCaptor.captureResponseInteraction(any<Response>(), body, elapsedTime) }
    }
}
